File: WeekStory7/PlacementMarathon.cpp
#include<iostream>
#include<vector>
using namespace std;
class PlacementMarathon {
    public:
    int compute(vector<int>& arr, int size, int gap, int prev, int cur, vector<vector<int>>& memo) {
        if(cur >= size) {
            return 0;
        }
        if(memo[prev+1][cur] != -1) {
            return memo[prev+1][cur];
        }
        int pick = 0;
        if(prev == -1 || (abs(arr[cur] - arr[prev]) == gap)) {
            pick = compute(arr, size, gap, cur, cur+1, memo) + 1;
        }
        int leave = compute(arr, size, gap, prev, cur+1, memo);
        return memo[prev+1][cur] = max(pick, leave);
    }
    int placementMarathon() {
        int size, gap;
        cout << "Enter number students and skill difference value(k):";
        cin >> size >> gap;
        vector<int> arr(size);
        cout << "Enter skill level of students:";
        for(int i = 0; i < size; i++) {
            cin >> arr[i];
        }
        vector<vector<int>> memo(size+1, vector<int>(size, -1));
        return compute(arr, size, gap, -1, 0, memo);
    }
};
int main() {
    PlacementMarathon* obj = new PlacementMarathon();
    cout << obj->placementMarathon() << endl;
    return 0;
}

File: week_stories_7/PlacementMarathon.java
import java.util.*;

public class PlacementMarathon {

    public int longestSequence(int[] skills, int n, int k, int prevIdx, int i, int[][] dp) {

        if (i >= n) {
            return 0;
        }

        if (dp[prevIdx + 1][i] != -1) {
            return dp[prevIdx + 1][i];
        }

        int take = 0;

        if (prevIdx == -1 || Math.abs(skills[i] - skills[prevIdx]) == k) {
            take = longestSequence(skills, n, k, i, i + 1, dp) + 1;
        }

        int skip = longestSequence(skills, n, k, prevIdx, i + 1, dp);

        return dp[prevIdx + 1][i] = Math.max(take, skip);
    }

    public int placementMarathon() {

        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number students and skill difference value(k): ");
        int n = sc.nextInt();
        int k = sc.nextInt();

        int[] skills = new int[n];

        System.out.println("Enter skill level of students:");

        for (int i = 0; i < n; i++) {
            skills[i] = sc.nextInt();
        }

        int[][] dp = new int[n + 1][n];

        for (int[] row : dp) {
            Arrays.fill(row, -1);
        }

        return longestSequence(skills, n, k, -1, 0, dp);
    }

    public static void main(String[] args) {

        PlacementMarathon obj = new PlacementMarathon();

        System.out.println(obj.placementMarathon());
    }
}
